### PR TITLE
Make OpenHands SDK condenser resume robust

### LINKED ISSUE #262: Make OpenHands SDK condenser resume robust
Track implementation and review for neulab/agent-data-protocol#261: Make OpenHands SDK condenser resume robust

Associated PR: https://github.com/neulab/agent-data-protocol/pull/261

Preserves source trajectory and row identifiers in OpenHands SDK condenser SFT records; supports source-row hash IDs for duplicate trajectory IDs during resumable generation; tolerates standardized rows that do not begin with a user text observation; expands ATIF schema support for the converter changes.

### PR DESCRIPTION
**Pull request by neubig**
## Summary\n- preserve source trajectory and row identifiers in OpenHands SDK condenser SFT records\n- support source-row hash IDs for duplicate trajectory IDs during resumable generation\n- tolerate standardized rows that do not begin with a user text observation\n- expand ATIF schema support for the converter changes\n\n## Verification\n- python -m py_compile agents/openhands_sdk/condensation_sft.py agents/openhands_sdk/std_to_sft.py schema/atif.py

Closes #262

### CODE REVIEW

File: schema/atif.py [hunk 1]
```diff
@@ -148,6 +141,14 @@ def validate_sequential_step_ids(self):
             raise ValueError(f"ATIF step_id values must be sequential starting at 1: {step_ids}")
         return self
 
+    @model_validator(mode="after")
+    def validate_observation_links(self):
+        # Full corpora can contain partial transcripts where an observation keeps a
+        # source_call_id from raw data but the corresponding tool-call turn is absent.
+        # Preserve the link when present; downstream converters can degrade unmatched
+        # observations to environment messages.
+        return self
```
[the remark this answers is not in the document]
Comment (neubig) [reply schema/atif.py:156]: Addressed in 2a7673f: removed the dead `validate_observation_links` model_validator entirely and moved its explanatory comment into the `ATIFTrajectory` class docstring, so the rationale is preserved without misleading readers with an empty validator.

File: agents/openhands_sdk/std_to_sft.py [hunk 2]
```diff
@@ -161,16 +193,89 @@ def stringify_value(value: Any, default: str = "") -> str:
...
+def infer_file_editor_command(command: str, kwargs: dict[str, Any]) -> str:
+    if command in {"append", "add"}:
+        return "insert"
+    if command in {"delete", "remove"}:
+        return "str_replace"
+    if "old_str" in kwargs or "new_str" in kwargs:
+        return "str_replace"
+    if "file_text" in kwargs:
+        return "create"
+    return "view"
```
[the remark this answers is not in the document]
Comment (neubig) [reply agents/openhands_sdk/std_to_sft.py:249]: Addressed in 2a7673f: `infer_file_editor_command` now logs a warning (`logger.warning`) describing the unrecognized command before falling back to `view`, so dropped mutations are surfaced rather than silent. I kept the `view` fallback rather than raising to avoid hard-failing conversion of full corpora on a single bad record, but the warning makes it observable. Also covered by new unit tests in `tests/test_openhands_sdk_std_to_sft_helpers.py`.

**Comment by neubig** [comment]
Addressed review feedback in 2a7673f:

1. Removed the dead `validate_observation_links` validator in `schema/atif.py` and moved its rationale into the `ATIFTrajectory` class docstring.
2. `infer_file_editor_command` now logs a warning when an unrecognized write-like command falls back to `view`, instead of silently dropping it.
3. Added unit tests in `tests/test_openhands_sdk_std_to_sft_helpers.py` covering the file-editor helpers and browser aliases.

Also fixed the failing `test (3.12)` CI: the `test_atif_schema_rejects_unknown_fields_and_bad_links` failure was caused by this PR's intentional relaxation of observation-link validation (updated the test to assert the relaxed behavior). The two `test_openhands_sdk_sft_conversion.py` failures were pre-existing — caused by `openhands-sdk` 1.29.0 wrapping the system prompt in `<SOUL>` tags (main CI passed with 1.28.1); made the assertion robust to that wrapper.

Both inline review threads have been resolved. Ready for another look.

File: tests/test_openhands_sdk_std_to_sft_helpers.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+    [
+        ("append", {}, "insert"),
+        ("add", {}, "insert"),
+        ("delete", {}, "str_replace"),
+        ("remove", {}, "str_replace"),
+        ("edit", {"old_str": "a", "new_str": "b"}, "str_replace"),
+        ("edit", {"new_str": "b"}, "str_replace"),
+        ("write", {"file_text": "hello"}, "create"),
+        ("view", {}, "view"),
+        ("unknown", {}, "view"),
```
[the remark this answers is not in the document]
Comment (neubig) [reply tests/test_openhands_sdk_std_to_sft_helpers.py]: Good catch — addressed in 42116a0. Removed the spurious `("view", {}, "view")` direct-call case from `test_infer_file_editor_command` (since `"view"` is a recognized command that never reaches `infer_file_editor_command` in production, the case just logged a spurious warning). To keep meaningful coverage of the recognized-command passthrough, added `test_finalize_file_editor_kwargs_recognized_command_skips_inference`, which routes `"view"` through the real entry point (`finalize_file_editor_kwargs`) and asserts no `Unrecognized ...` warning is emitted. The fallback path itself remains covered by the `("unknown", {}, "view")` parametrize case plus `test_infer_file_editor_command_warns_on_unknown`.

**Comment by neubig** [comment]
Addressed the test-quality suggestion in 42116a0: dropped the spurious `("view", {}, "view")` direct-call case from `test_infer_file_editor_command` (since "view" is a recognized command that never reaches `infer_file_editor_command` in production) and added `test_finalize_file_editor_kwargs_recognized_command_skips_inference` to exercise the recognized-command passthrough via the real entry point, asserting no spurious warning is emitted. CI is green (1106 passed, 1 skipped). Review thread resolved. Ready for another look.

File: tests/test_openhands_sdk_std_to_sft_helpers.py [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+    "command,kwargs,expected",
+    [
+        ("append", {}, "insert"),
+        ("add", {}, "insert"),
+        ("delete", {}, "str_replace"),
+        ("remove", {}, "str_replace"),
+        ("edit", {"old_str": "a", "new_str": "b"}, "str_replace"),
+        ("edit", {"new_str": "b"}, "str_replace"),
+        ("write", {"file_text": "hello"}, "create"),
+        ("unknown", {}, "view"),
```
[the remark this answers is not in the document]
Comment (neubig) [reply tests/test_openhands_sdk_std_to_sft_helpers.py]: Addressed in 4689445. Dropped the redundant `("unknown", {}, "view")` parametrized case entirely — the no-match fallback (-> "view" + warning) is fully covered by the dedicated `test_infer_file_editor_command_warns_on_unknown` with `caplog`, so this case no longer emits a spurious warning during the parametrized run. The remaining parametrize cases still cover the recognized-name -> edit-op mapping (append/add -> insert, delete/remove -> str_replace, edit -> str_replace, write -> create).

File: agents/openhands_sdk/std_to_sft.py [hunk 5]
```diff
@@ -161,16 +196,94 @@ def stringify_value(value: Any, default: str = "") -> str:
...
+        command,
+    )
+    return "view"
+
+
+def finalize_file_editor_kwargs(kwargs: dict[str, Any]) -> dict[str, Any]:
+    command = stringify_value(kwargs.get("command"))
+    if command not in FILE_EDITOR_COMMANDS:
+        command = infer_file_editor_command(command, kwargs)
+    kwargs["command"] = command
```
[the remark this answers is not in the document]
Comment (neubig) [reply agents/openhands_sdk/std_to_sft.py:256]: Addressed in 4689445 exactly as suggested. `finalize_file_editor_kwargs` now opens with `kwargs = dict(kwargs)` and operates on the local copy, so the caller's dict is never mutated — it's now a pure function. The returned value is identical. Added `test_finalize_file_editor_kwargs_does_not_mutate_input` to lock in the non-mutation behavior.

**Comment by neubig** [comment]
Addressed both 🟡 suggestions in 4689445:

1. Dropped the redundant `("unknown", {}, "view")` parametrized case from `test_infer_file_editor_command` — the no-match fallback (-> "view" + warning) is fully covered by the dedicated `test_infer_file_editor_command_warns_on_unknown` with `caplog`, so the parametrized run no longer emits a spurious warning. Recognized-name -> edit-op mapping still covered.
2. `finalize_file_editor_kwargs` now opens with `kwargs = dict(kwargs)` and operates on the local copy, making it a pure function (no input mutation). Returned value identical. Added `test_finalize_file_editor_kwargs_does_not_mutate_input`.

CI green: `test (3.12)` 1106 passed, 1 skipped; pre-commit, check_dataset_metadata, check-pr-artifacts all pass. Both review threads resolved. Ready for another look.

File: tests/test_openhands_sdk_std_to_sft_helpers.py [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+def test_normalize_direct_file_editor_kwargs(function_name, kwargs, expected_command):
+    finalized = normalize_direct_file_editor_kwargs(function_name, kwargs)
+    assert finalized["command"] == expected_command
+
+
+def test_normalize_direct_file_editor_kwargs_unknown_function_warns():
+    # An unrecognized write-like command without old_str/new_str/file_text should
+    # surface a warning and degrade to "view" rather than silently succeeding.
+    finalized = normalize_direct_file_editor_kwargs("weird", {})
+    assert finalized["command"] == "view"
```
[the remark this answers is not in the document]
Comment (neubig) [reply tests/test_openhands_sdk_std_to_sft_helpers.py]: Addressed in 9cada16, exactly as suggested. `test_normalize_direct_file_editor_kwargs_unknown_function_warns` now takes the `caplog` fixture, wraps the call in `caplog.at_level(logging.WARNING, logger=...)`, and asserts that an "Unrecognized file-editor command" warning is logged — matching the pattern used by the sibling warning tests. I verified the scenario genuinely emits the warning (the unrecognized function name flows through `finalize_file_editor_kwargs` -> `infer_file_editor_command` and logs it), so the assertion is real. CI green: 1106 passed, 1 skipped.

**Comment by neubig** [comment]
Addressed the 🟡 suggestion in 9cada16: `test_normalize_direct_file_editor_kwargs_unknown_function_warns` now uses the `caplog` fixture and asserts the "Unrecognized file-editor command" warning is logged at WARNING level, matching the sibling warning-test pattern. I confirmed the scenario genuinely emits the warning (unrecognized function -> finalize_file_editor_kwargs -> infer_file_editor_command), so the assertion is real.

Note: the initial push of 9cada16 did not trigger GitHub Actions (webhook miss — zero check-runs for that SHA), so I pushed an empty trigger commit 1651dda to re-fire the `synchronize` event. CI now green: `test (3.12)` 1106 passed, 1 skipped; pre-commit, check_dataset_metadata, check-pr-artifacts all pass. Review thread resolved. Ready for another look.

File: agents/openhands_sdk/std_to_sft.py [hunk 7]
```diff
@@ -224,24 +339,94 @@ def create(
     )
 
 
+def make_fallback_tool(name: str) -> type[ToolDefinition]:
+    description = f"Undeclared dataset tool {name}."
+
+    def create(
+        cls,
+        conv_state=None,  # noqa: ARG001
+        _description=description,
```
[the remark this answers is not in the document]
Comment (neubig) [reply agents/openhands_sdk/std_to_sft.py:338]: Addressed in 67efd97 exactly as suggested. Added `**params,  # noqa: ARG001` to `make_fallback_tool`'s `create` closure, matching the sibling `make_metadata_tool` pattern (line 324) for consistency and forward-compatibility against the SDK framework passing extra kwargs.

File: agents/openhands_sdk/condensation_sft.py [hunk 8]
```diff
@@ -227,20 +256,30 @@ def append_standardized_events_with_condensation(
...
+                llm_message=Message(
+                    role="user",
+                    content=[
+                        TextContent(text="Continue the task from the current workspace state.")
+                    ],
+                ),
+            )
         )
-    )
+        index = 0
```
[the remark this answers is not in the document]
Comment (neubig) [reply agents/openhands_sdk/condensation_sft.py:282]: Addressed in 67efd97 via the lower-risk doc option (no runtime behavior change). Added a docstring to `append_standardized_events_with_condensation` explaining that `start_index` is respected only when the first trajectory event is a user `TextObservation`, and that on the fallback path (no leading user message to consume) it is intentionally reset to 0 so no trajectory content is skipped. Also added a short inline comment at the override site pointing back to the docstring. The parameter is retained since the only caller passes `start_index=1`.

**Comment by neubig** [comment]
Addressed both 🟡 suggestions in 67efd97:

1. `make_fallback_tool`'s `create` closure now accepts `**params,  # noqa: ARG001`, matching the sibling `make_metadata_tool` pattern for consistency and forward-compatibility (avoids a potential TypeError if the SDK framework passes extra kwargs).
2. Documented the `start_index` override in `append_standardized_events_with_condensation` via a docstring (lower-risk option, no runtime behavior change): `start_index` is respected only when the first trajectory event is a user `TextObservation`; on the fallback path it is intentionally reset to 0 so no trajectory content is skipped. Added a short inline comment at the override site pointing to the docstring.

No runtime behavior changed; the caplog assertion from the prior fix is intact. CI green: `test (3.12)` 1106 passed, 1 skipped; pre-commit, check_dataset_metadata, check-pr-artifacts all pass. Both review threads resolved.

Note: the push of 67efd97 did not trigger GitHub Actions (webhook miss, recurring for this branch — zero check-runs for that SHA), so I pushed an empty trigger commit 4e39d23 to re-fire the `synchronize` event. The code fix lives in 67efd97; 4e39d23 contains no content changes. Ready for another look.

File: tests/test_openhands_sdk_std_to_sft_helpers.py [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+        ("get", {}, "browser_get_content", {"extract_links": False, "start_from_char": 0}),
+        ("click", {"index": 7}, "browser_click", {"index": 7, "new_tab": False}),
+        ("click", {"bid": "el[3]"}, "browser_click", {"index": 3, "new_tab": False}),
+        ("type", {"index": 2, "text": "hi"}, "browser_type", {"index": 2, "text": "hi"}),
+        ("fill", {"index": 1, "value": "v"}, "browser_type", {"index": 1, "text": "v"}),
+        ("press", {"index": 0, "key": "Enter"}, "browser_type", {"index": 0, "text": "Enter"}),
+        ("clear", {"index": 4}, "browser_type", {"index": 4, "text": ""}),
+        ("scroll", {"direction": "up"}, "browser_scroll", {"direction": "up"}),
+        ("scroll", {"delta_y": -5}, "browser_scroll", {"direction": "up"}),
+        ("upload_file", {}, "browser_get_state", {"include_screenshot": False}),
```
[the remark this answers is not in the document]
Comment (neubig) [reply tests/test_openhands_sdk_std_to_sft_helpers.py]: Addressed in 866c896. Added three `select_option` cases to `test_map_browser_action`'s parametrize list, genuinely exercising the branch's non-trivial logic: a single-value `options` case, a list `options` case (exercising the `'\n'.join` path -> `"x\ny"`), and a `value` fallback case. All pass locally and in CI (1109 passed).

**Comment by neubig** [comment]
Addressed both final items in 866c896:

1. **`select_option` coverage**: Added three cases to `test_map_browser_action`'s parametrize list, genuinely exercising the branch's non-trivial logic — a single-value `options` case, a list `options` case (exercising the `'\n'.join` path → `"x\ny"`), and a `value` fallback case. (Per the constraint, no no-op cases.)
2. **Dead-code fallback**: Replaced `return BROWSER_TOOL_ALIASES[function_name], kwargs` in `map_browser_action` with `raise AssertionError(f"unhandled browser action: {function_name!r}")`. I verified all 13 `BROWSER_TOOL_ALIASES` keys have explicit branches above (and `map_browser_action` is only called for known aliases via `should_map_to_browser_action`), so this never fires for valid input today but surfaces a future unhandled alias immediately.

All previous fixes intact (caplog assertion, `**params`, `start_index` doc note). CI green: `test (3.12)` 1109 passed, 1 skipped; pre-commit, check_dataset_metadata, check-pr-artifacts all pass. All review threads resolved. Ready for another look.

File: agents/openhands_sdk/condensation_sft.py [hunk 10]
```diff
@@ -46,6 +47,18 @@
...
+def source_row_id_from_line(line: str, trajectory_id: str) -> str:
+    row = json.loads(line)
+    canonical = json.dumps(
+        row,
+        sort_keys=True,
+        separators=(",", ":"),
+        ensure_ascii=False,
+    )
+    digest = hashlib.sha256(canonical.encode("utf-8")).hexdigest()[:16]
+    return f"{trajectory_id}__row_{digest}"
```
[the remark this answers is not in the document]
Comment (neubig) [reply agents/openhands_sdk/condensation_sft.py:59]: Addressed in b212ed6. Added a dedicated `tests/test_openhands_sdk_condensation_sft.py` covering all three requested cases plus more:
- determinism (same input → same output)
- disambiguation (different content → different ID; different trajectory_id → different ID)
- documented format `<trajectory_id>__row_<16-hex-chars>` (asserts the f-string and 16-char digest length)
- canonical key ordering (reordered raw line → same ID) and an empty-object edge case
- a parametrized test pinning exact digest values (e.g. `traj_123__row_43258cff783fe703`, `traj_123__row_44136fa355b3678a` for `{}`) so accidental changes to the hashing scheme are caught.

File: tests/test_openhands_sdk_std_to_sft_helpers.py [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+        ("select_option", {"index": 2, "options": "a"}, "browser_type", {"index": 2, "text": "a"}),
+        (
+            "select_option",
+            {"index": 5, "options": ["x", "y"]},
+            "browser_type",
+            {"index": 5, "text": "x\ny"},
+        ),
+        ("select_option", {"index": 1, "value": "v"}, "browser_type", {"index": 1, "text": "v"}),
+        ("scroll", {"direction": "up"}, "browser_scroll", {"direction": "up"}),
+        ("scroll", {"delta_y": -5}, "browser_scroll", {"direction": "up"}),
```
[the remark this answers is not in the document]
Comment (neubig) [reply tests/test_openhands_sdk_std_to_sft_helpers.py]: Addressed in b212ed6. Extended `test_map_browser_action`'s scroll parametrize list with exactly the suggested cases plus an explicit `direction: 'down'`:
```python
("scroll", {"delta_y": 5}, "browser_scroll", {"direction": "down"}),
("scroll", {"direction": "down"}, "browser_scroll", {"direction": "down"}),
("scroll", {}, "browser_scroll", {"direction": "down"}),
```
The positive `delta_y` case exercises the `delta >= 0` → 'down' branch, and the no-arg case exercises the default `direction = 'down'` fallback.

**Comment by neubig** [comment]
Addressed both final test-coverage suggestions in b212ed6:

1. **`source_row_id_from_line` tests**: Added a dedicated `tests/test_openhands_sdk_condensation_sft.py` covering all requested cases and more — determinism (same input → same output), disambiguation (different content → different ID; different trajectory_id → different ID), the documented format `<trajectory_id>__row_<16-hex-chars>` (asserting the f-string and 16-char digest length), canonical key ordering, an empty-object edge case, and a parametrized test pinning exact digest values (e.g. `traj_123__row_43258cff783fe703`) so accidental hashing-scheme changes are caught.

2. **`scroll` 'down' coverage**: Extended `test_map_browser_action`'s scroll parametrize list with the suggested positive-`delta_y` and no-arg cases plus an explicit `direction: 'down'` case — exercising the `delta >= 0` → 'down' branch and the default fallback.

All previous fixes intact (caplog assertion, `**params`, `start_index` doc note, `select_option` cases, `AssertionError` fallback). CI green: `test (3.12)` 1122 passed, 1 skipped; pre-commit, check_dataset_metadata, check-pr-artifacts all pass. Both review threads resolved. Ready for another look.
